checklist/src/main/scala/checklist/Rule.scala

```diff
@@ -4,20 +4,36 @@ import cats.{Applicative, Traverse}
 import cats.data.Ior
 import cats.instances.all._
 import cats.syntax.all._
-import monocle.{PLens, Lens}
+import monocle.PLens
 import scala.language.higherKinds
 import scala.util.matching.Regex
-import Message.{errors, warnings}
+import Message.errors
+import cats.data.NonEmptyList
 
 sealed abstract class Rule[A, B] {
   def apply(value: A): Checked[B]
 
   def map[C](func: B => C): Rule[A, C] =
     Rule.pure(value => this(value) map func)
 
+  def emap[C](func: B => Checked[C]): Rule[A, C] =
+    Rule.pure(value => this(value) flatMap func)
+
+  def recover(func: Messages => Checked[B]): Rule[A, B] =
+    Rule.pure(value => this(value).fold(func, Ior.right, Ior.both))
+
+  def mapMessages(func: Messages => Messages): Rule[A, B] =
+    Rule.pure(value => this(value).fold(func andThen Ior.left, Ior.right, (msgs, r) => Ior.both(func(msgs), r)))
+
+  def mapEachMessage(func: Message => Message): Rule[A, B] =
+    mapMessages(_.map(func))
+
   def contramap[C](func: C => A): Rule[C, B] =
     Rule.pure(value => this(func(value)))
 
+  def contramapPath[C, D: PathPrefix](path: D, func: C => A): Rule[C, B] =
```

Comment:
I'm wondering if this method ought to have two lists of parameters.

Reply:
I'm inclined to agree. At the time of writing, all my contramap functions were so simple that I couldn't think of a reason to separate them.  I doubt that'll always be the case though.

Reply:
Can you tweak the PR for this please? Happy to merge tonight!

Reply:
Done!